src/SegmentedArray.chpl

```diff
@@ -75,16 +70,228 @@ module SegmentedArray {
+                throw new owned OutOfBoundsError();
+            }
+            // Early return for zero-length result
+            if (size == 0) || (slice.size == 0) {
+                return (makeDistArray(0, int), makeDistArray(0, values.etype));
+            }
+
+            ref sa = segments.a;
+            var start = sa[slice.low];
+            // end index
+            var end: int = if slice.high == segments.aD.high then values.size-1 else sa[slice.high:int+1]-1;
+
+            // Segment offsets of the new slice
+            var newSegs = makeDistArray(slice.size, int);
+            // TODO - double check this aggregation
+            forall (i, ns) in zip(newSegs.domain, newSegs) with (var agg = newSrcAggregator(int)) {
+                agg.copy(ns, sa[slice.low + i]);
+            }
+
+            // re-zero segments
+            newSegs -= start;
+            
+            var newVals = makeDistArray(end - start + 1, values.etype);
+            ref va = values.a;
+            forall (i, nv) in zip(newVals.domain, newVals) with (var agg = newSrcAggregator(values.etype)) {
+                agg.copy(nv, va[start + i]);
+            }
+            return (newSegs, newVals);
+        }
+
+        /* Gather segments by index. Returns arrays for the segments and values.*/
+        proc this(iv: [?D] ?t, smallIdx=false) throws where t == int || t == uint {
+            use ChplConfig;
+
+            // Early return for zero-length result
+            if (D.size == 0) {
+                return (makeDistArray(0, int), makeDistArray(0, values.etype));
+            }
+            // Check all indices within bounds
+            var ivMin = min reduce iv;
+            var ivMax = max reduce iv;
+            if (ivMin < 0) || (ivMax >= segments.size) {
+                saLogger.error(getModuleName(),getRoutineName(),getLineNumber(),
+                                    "Array out of bounds");
+                throw new owned OutOfBoundsError();
+            }
+            saLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                                                    "Computing lengths and offsets");
+            var t1 = getCurrentTime();
+            ref oa = segments.a;
+            const low = segments.aD.low, high = segments.aD.high;
+
+            // Gather the right and left boundaries of the indexed strings
+            // NOTE: cannot compute lengths inside forall because agg.copy will
+            // experience race condition with loop-private variable
+            var right: [D] int, left: [D] int;
+            forall (r, l, idx) in zip(right, left, iv) with (var agg = newSrcAggregator(int)) {
+                if (idx == high) {
+                    agg.copy(r, values.size);
+                } else {
+                    agg.copy(r, oa[idx:int+1]);
+                }
+                agg.copy(l, oa[idx:int]);
+            }
+
+            // Lengths of segments including null bytes
```

Comment:
this comment should be updated to remove `including null bytes`

Reply:
Done